(MODULES-2676) Support MySQL 5.7.6+

elconas opened the pull request:
Fixes Password Handling for MySQL 5.7.6++
Fixes mysql_install_db for MySQL 5.7.6++
Fixes removed myisam-recove on MySQL 5.7.6++
Added Support for mysqld_version 
Added Provider to create data_directory (with version detection)

MySQL 5.7.6 changed password related SQL commands which needed fixing (done in a prior pull request already), however introducing explicit versioning seems required to support MySQL 5.7 as there are many more incompatible changes (e.g. init db, see below). Also the provider can be used to have better "rspec" tests.

Also MySQL 5.7.6 introduces mysql --initialize-insecure instead of mysql_install_db. To support this without "exec | unless mysqld -V | grep 5.7...." a provider has been written. Also the mysqld_version attribute was added to the base provider to support general version dependent code. 

MySQL 5.7.6 also removed myisam-recover, thus this parameter needs to go to version specific [mysqld-5.X] parameter sections. 

This change has been tested on CentOS 6.7 with MySQL Community Editions: 

  - MySQL Community 5.5 (5.5.46)
  - MySQL Community 5.6 (5.6.27)
  - MySQL Community 5.7 (5.7.8)

Regards, 
Robert

DavidS commented on lib/puppet/provider/mysql.rb:
```diff
@@ -13,7 +14,22 @@ def self.defaults_file
+    # note: be prepared for '5.7.6-rc-log' etc results
+    #       versioncmp detects 5.7.6-log to be newer then 5.7.6
+    #       this is why we need the trimming.
+    return @mysqld_version_string.scan(/\d+\.\d+\.\d+/).first unless @mysqld_version_string.nil?
```
Shouldn't this regex (and the one below) be anchored with `^`, or at least have an additional space in front to not accidentally match something else?

DavidS commented on lib/puppet/type/mysql_datadir.rb:
```diff
@@ -0,0 +1,31 @@
+Puppet::Type.newtype(:mysql_datadir) do
+  @doc = 'Manage MySQL datadirs with mysql_install_db OR msqld (5.7.6 and above).'
```
s/msqld/mysqld/

DavidS commented on lib/puppet/type/mysql_datadir.rb:
```diff
@@ -0,0 +1,31 @@
+
+  newparam(:user) do
+    desc 'The user for the directory default mysql (name, not uid).'
+    newvalues(/\D+/)
```
user names can contain numbers

DavidS commented on spec/unit/puppet/provider/mysql_user/mysql_spec.rb:
```diff
@@ -85,6 +98,38 @@
+    it 'detects < 5.7.6 for 5.7.5-log' do
+       provider.class.instance_variable_set(:@mysqld_version_string, 'mysqld  Ver 5.7.5-log for Linux on x86_64 (MySQL Community Server (GPL))')
+       expect(Puppet::Util::Package.versioncmp(provider.mysqld_version, '5.7.6')).to be < 0
+    end
```
You could compress this whole block with

```ruby
[
  [ 'mysqld  Ver 5.7.6 for Linux on x86_64 (MySQL Community Server (GPL))', '5.7.6 ]
  # ...
].each do |dash_v, version|
  it "detects '#{dash_v}' as '#{version}'" do
    provider.class.instance_variable_set(:@mysqld_version_string, dash_v)
    expect(provider.mysqld_version).to eq version
  end
end
```

igalic replied:
:purple_heart:

DavidS commented on spec/unit/puppet/provider/mysql_user/mysql_spec.rb:
```diff
@@ -103,12 +148,19 @@
+      provider.expects(:mysql).with([defaults_file, '-e', "ALTER USER 'joe'@'localhost' IDENTIFIED WITH mysql_native_password AS '*6C8989366EAF75BB670AD8EA7A7FC1176A95CEF5'"]).returns('0')
+
+      provider.expects(:password_hash).returns('*6C8989366EAF75BB670AD8EA7A7FC1176A95CEF5')
+      provider.password_hash=('*6C8989366EAF75BB670AD8EA7A7FC1176A95CEF5')
```
When you're already in there, at the top of the tests, there is a `newhash` defined which has this value, I think this should use that then?

DavidS commented:
Hi, thank you for your contribution. I've added only a few minor comments to the code. Additionally it requires a few sentences in the README's reference section explaining the new type.

DavidS commented:
PS: @igalic , please also have a look at this!

elconas commented:
Hello, fixed according to your feedback. 

Also I fixed MariaDB 10 regression (which is actually 5.6, but is detected as > 5.7.6). Now I added "flavor" detection and made the 5.7.6 changes very closely coupled with MySQL type.

DavidS commented on README.md:
```diff
@@ -792,6 +795,17 @@ The name of the MySQL plugin to manage.
+Initializes the MySQL data directory with version specific code. Pre MySQL 5.7.6
+it uses mysql_install_db. After MySQL 5.7.6 it uses mysqld --initialize-insecure.
+
+Insecure initialization is needed, as mysqld version 5.7 indroduced "secure by default" mode.
```
s/indroduced/introduced

DavidS commented on lib/puppet/provider/mysql_datadir/mysql.rb:
```diff
@@ -0,0 +1,65 @@
+
+  def create
+    name                     = @resource[:name]
+    insecure                 = @resource.value(:insecure) || :true
```
Please use just plain booleans here instead of `:true`/`:false`. Puppet core has even a separate boolean property: https://docs.puppetlabs.com/guides/custom_types.html#boolean-parameters

DavidS commented on lib/puppet/provider/mysql_datadir/mysql.rb:
```diff
@@ -0,0 +1,65 @@
+      end
+    end
+
+   exists? ? (return true) : (return false)
```
Since `exists?` already returns a bool, you can just say `exists?` instead of funging around with the ternary operator here.

DavidS commented:
Since this is a larger change, I'd be very thankful if someone else, like @igalic , @roman-mueller or @mcanevet , could have a look.

Meanwhile, it'd be great if you could rebase/squash this into a single commit to avoid cluttering the git history. (See https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#Squashing-Commits ).

DavidS commented:
PS: Adding `(MODULES-2676)` into the commit's title will automatically link this back to the JIRA ticket.

mcanevet commented:
This is a scary one... @elconas would it be possible to merge this step by step? For example, a first PR that replaces `Exec[mysql_install_db]` by the `mysql_datadir` resource. Then it would be easier to add MySQL 5.7.6++ support.

elconas commented:
Hi, this is difficult - from my perspective - as I need Versioning (5.5.X) and Typing (mysql/percona/mariadb) support for replacing the Exec[mysql_install_db]. So if I have Versioning and Typing, adding the two different version specific SQL statements in the user management section of the manifest is kind of trivial. 

This makes the replace Exec containing 90% of the change, making this change not really splittable.

elconas commented:
Did in same branch, to keep the pull request and the history: 

git rebase -i 2a8693482f962b36024e9c70a459de8e9dc65eb6
git push origin +fix_mysql576_pull

DavidS commented:
I'm now running it locally through some tests and will merge unless something pops up.

DavidS commented:
@elconas the tests fail with `Error: Could not find a suitable provider for mysql_datadir` as `mysqld` is not a command on the path on Centos 6.4. Instead it is placed into `/usr/libexec`. Since the agent doesn't look for it there, it marks the provider as invalid and fails.

mentat commented:
@elconas can you rebase this and ensure the tests pass?

igalic commented:
@mentat the centos7 tests can, for now, be ignored.

mentat commented:
Hi @elconas could you rebase this one so we can get it merged?

DavidS commented:
I've rebased this to #786 and addressed the merge conflicts there.

DavidS commented:
and it was merged. Thanks for all your work!

mentat commented:
Hi @elconas it seems this PR fails for redhat 6 & 7 but works on Ubuntu.  Do you have any ideas how to improve the compatibility?

elconas commented:
Hi, Fixed in https://github.com/puppetlabs/puppetlabs-mysql/pull/789
